Allow changing level or tilt while blind is moving

sleiner opened the pull request:
Addresses #1506.

Test setup:
- HMIP-BBL (firmware version 1.10.6)
- RaspberryMatic 3.75.7.20240420
- custom_homematic [`53f70f7`](https://github.com/danielperna84/custom_homematic/commit/53f70f7e70763969ef1518503c4ecac167654ae8)

sleiner committed:
Allow changing level or tilt while blind is moving

sleiner commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -316,24 +364,32 @@ async def close(self, collector: CallParameterCollector | None = None) -> None:
+        """Stop the device if in motion. Use only when command_processing_lock is held."""
+        await self.stop(collector=collector)
 
     def is_state_change(self, **kwargs: Any) -> bool:
```
This could be made aware of target positions

sleiner replied:
Ich habe das mal versucht - so richtig glücklich war ich mit dem Ergebnis nicht. Das Problem ist, dass sowohl die einzelnen Methoden (wie `open`, `close`, `open_tilt`, `vent`, ...) als auch `is_state_change` unabhängig voneinander bestimmen, wie Target Level (und Target Tilt Level) abhängig von der auszuführenden Aktion zu setzen sind. Wenn hier noch eine Überprüfung von `target_level` und `target_tilt_level` dazukommt, wird es schnell sehr unübersichtlich.

Aus meiner Sicht wäre es sinnvoll, das Ableiten der Ziellevel und die Überprüfung, ob für gegebene Ziellevel ein neues Kommando gesendet werden muss, klar zu separieren. Dann kann man das Ableiten der Ziellevel auch an einer Stelle tun. Das wäre aber auch in nem eigenen PR ganz gut aufgehoben - was meinst du, @SukramJ?

SukramJ replied:
Hört sich gut am

sleiner commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -279,22 +294,56 @@ async def _set_level(
+                level=_level, tilt_level=_tilt_level, collector=collector
+            )
+
+    @bind_collector(use_command_queue=False)
```
I moved the collectors further "down" the call graph. The reason is that the `_target` properties should be changed before the lock is released, so that no race conditions can occur.

sleiner reviewed (commented):
@SukramJ I will be offline for the rest of the week, but I wanted to get a first draft out there for you to review the overall concept.

SukramJ commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -279,22 +294,56 @@ async def _set_level(
+            )
+
+    @bind_collector(use_command_queue=False)
+    async def _transmit_position(
```
Bitte umbennen auf `_send_level`

sleiner replied:
Erledigt: 11c1962d52e74e7b247e2527452b3cfbf70ec244

SukramJ commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -279,22 +294,56 @@ async def _set_level(
+        tilt_level: float,
+        collector: CallParameterCollector | None = None,
+    ) -> None:
+        """Transmit a new target position to the device."""
```
position -> level

sleiner replied:
Erledigt: 11c1962d52e74e7b247e2527452b3cfbf70ec244

SukramJ commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -316,24 +364,32 @@ async def close(self, collector: CallParameterCollector | None = None) -> None:
+        async with self._command_processing_lock:
+            await self._stop(collector=collector)
+
     @bind_collector(use_command_queue=False)
```
`use_command_queue=False `ist der default und muss nicht gesetzt werden. 
Alternativ `wait_for_callback=60, use_command_queue=True `verenden.

sleiner replied:
Erledigt: 055de8f23f7794abfbca28733685b937f7df09e6

SukramJ commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -279,22 +294,56 @@ async def _set_level(
+                level=_level, tilt_level=_tilt_level, collector=collector
+            )
+
+    @bind_collector(use_command_queue=False)
```
use_command_queue=False ist der default und muss nicht gesetzt werden.

sleiner replied:
Erledigt: 055de8f23f7794abfbca28733685b937f7df09e6

Commit:
Fix empty_queue

sleiner committed:
Rename _transmit_position to _send_level

sleiner committed:
Do not send use_command_queue explicitly

sleiner committed:
Merge branch 'devel' into fix/new-blind-position-while-moving

sleiner committed:
Reproduce race condition better in test case

Previously, the test case would succeed even when a command queue was
being used. Since that leads to bugs on a real CCU (where the time for
sending a command is not negligible), this commit introduces a similar
communication delay to the test.

At the same time, this allows us to reduce the number of iterations that
this test needs to run in order to be reasonably sure that a race
condition is absent.

SukramJ commented on hahomematic/platforms/custom/cover.py:
```diff
@@ -316,24 +362,32 @@ async def close(self, collector: CallParameterCollector | None = None) -> None:
+    async def _stop(self, collector: CallParameterCollector | None = None) -> None:
+        """Stop the device if in motion. Do only call with _command_processing_lock held."""
+        self.central.command_queue_handler.empty_queue(address=self._channel_address)
+        await super().stop()
```
Sollte sein:
`await self._stop(collector=collector)`

sleiner replied:
Ich vermute, es ging dir darum, `collector` weiterzugeben - richtig? Falls ja: 52940b6ffa19118aaf87aa54ea23904656e74508

sleiner committed:
Pass collector argument

sleiner committed:
Merge branch 'devel' into fix/new-blind-position-while-moving
